### DISCUSSION TITLE
Parallel pybamm

### DISCUSSION THREAD

**Discussion by akariwrynn** [pybamm-team/PyBaMM#3334] [Q&A]
Hi,
I hope everyone is fine. I have asked this question almost 2 months ago but my supervisior asked me to ask one more time. So, I know that pybamm cannot be parallelized currently. Therefore, I check if it is single instance code. When I run 2 simulations at the same time, each simulation becomes slower than a single simulation. Specifically, a single simulation runs for 77 seconds but two simulations at the same time runs for 106 second each. Therefore, we think that it is serialized and single instance. Maybe this is the reason that we cannot parallelized pybamm. So, my question is what is the real reason that we cannot parallelized? Because we are eager to look to the code and consult a profesor from CS department in order to both contribute this amazing community and also continue our interesting project. 

Thanks a lot,
Eren

**Comment by TomTranter** [comment]
hi @akariwrynn,
Thanks for the question and sorry it didn't get answered 2 months ago. Did you raise a discussion back then? We do have some limited support for parallel processing in PyBaMM which can be used via input parameters and the ScipySolver. Here is an example:
```
import pybamm
import numpy as np

def run():
    model = pybamm.lithium_ion.SPM()
    param = pybamm.ParameterValues("Chen2020")
    input_var = "Current function [A]"
    param.update({input_var: "[input]"})
    solver = pybamm.ScipySolver(rtol=1e-8, atol=1e-8)
    t_eval = np.linspace(0, 10, 100)
    ninputs = 4
    inputs_list = [{input_var: 5 * (1-(i + 1)/10)} for i in range(ninputs)]
    sim = pybamm.Simulation(model=model, parameter_values=param, solver=solver)
    solutions = sim.solve(t_eval, inputs=inputs_list, nproc=4)
    return solutions
    
if __name__ == "__main__":
    solutions = run()
    plot = pybamm.QuickPlot(solutions=solutions)
    plot.dynamic_plot()
```
Unfortunately there are some limitations where ScipySolver is only for ODEs and won't work for DFN models and also not every parameter can be changed to an input. There is also another repository for running pack simulations that I have made called liionpack https://github.com/pybamm-team/liionpack and within this framework I run the cell models in parallel and make use of the casadi solver. This could be easily tweaked I think to simply run generic models in parallel without the pack configuration part and would not be limited to ODEs. However, that code could still be improved and I hit performance issues above a few thousand parallel models. 

If you are keen to contribute to this area of PyBaMM we would welcome a discussion with you and any of your colleagues. CC @martinjrobins

**Reply by akariwrynn** [reply]
Thanks a lot for the answer and I think @martinjrobins answer my issue but as I said I wanted to ask one more time. Sorry for bothering too much. For our project we are trying to fit some parameters such as cathode thickness and positive particle radius for a battery measured in different temperatures. Our goal is to parallelize several temperatures and several ac amplitudes. Therefore, we need several simulations running at the same time with a good speed that we will not wait for days. Do you suggest the mentioned liionpack package or dive into to the code (DFN model) and try to figure out what can we do? If your answer is the second choice, can you suggest any reason why we cannot parallelize pybamm?

Thanks a lot and take care,
Eren

**Reply by TomTranter** [reply]
How long does one simulation take you?

**Reply by akariwrynn** [reply]
Sorry for late response. I did a quick run for one simulation and it takes 561 seconds. However, generally when we run the fit algorithm, to find the best values for the parameters, it takes 2 or 3 days at most. However, we want to increase the number of parameters and number of temperatures. Before doing so we want to parallelize the code. Here is an example code that we done for one simulation run:

```ruby

import pybamm
import numpy as np
import time
pointsperperiod=2048
period=10
fmin=-3
fmax=3
fspace=80
period=10

start1=time.time()
amplitude=0.01
parameters={"Positive particle radius":5.22*10**-6,"Positive electrode thickness":75.6*10**-6,"Negative particle radius":5.8*10**-6,"Negative electrode thickness":8.52*10**-5,"Separator Bruggeman coefficient":1.5,"Positive electrode diffusivity":4*10**-15,"Negative electrode diffusivity":3.3*10**-14}
model=pybamm.lithium_ion.DFN()
mychemistry=pybamm.parameter_sets.Chen2020
freq=np.logspace(-fmin,fmax,fspace)
for i in freq:
    parameters=dict(parameters)
    params = pybamm.ParameterValues(mychemistry)
    current=lambda t: amplitude * pybamm.sin(2 * np.pi * i * t)
    parameters['Current function [A]'] = current
    time_array= np.linspace(0, period/i, pointsperperiod*period)

    solver=pybamm.CasadiSolver("safe")
    sim = pybamm.Simulation(model, parameter_values=params,solver=solver)
    sim.solve(t_eval=time_array)
    solution = sim.solution
    V = solution["Terminal voltage [V]"]

end1=time.time()
print(f"Without pool:{end1-start1}")

```

**Reply by TomTranter** [reply]
Well I don't know what you've tried so far in terms of parallelization but it's possible. If you like I can take a look at the full script. Are you on the pybamm slack channel?
